proposed/resource-limit.md

```diff
@@ -0,0 +1,843 @@
+
+#### Pipelines
+
+TBD
+
+## Proof-of-concepts
+
+### Customized resource limiter based on ambient state that won't ship in box
+
+e.g. The limiter fails when CPU usage is above 90%
+
+```c#
+class CPUUsageLimiter : ResourceLimiter
+{
+    private PerformanceCounter _cpuUsageCounter = new PerformanceCounter("Processor", "% Processor Time", "_Total");
+    private float _threshold;
+
+    public CPUUsageLimiter(float threshold)
+    {
+        _threshold = threshold;
+    }
+
+    public long EstimatedCount { get; } => (long)_cpuUsageCounter.NextValue();
+
+    public Resource Acquire(long requestedCount)
+    {
+        if (cpuCounter.NextValue() > _threshold)
+        {
+            return Resource.FailNoopResource;
+        }
+        return Resource.SuccessNoopResource;
+    }
+
+    public ValueTask<Resource> AcquireAsync(long requestedCount, CancellationToken cancellationToken = default)
+    {
+        while(true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            if (cpuCounter.NextValue() <= _threshold)
+            {
+                return Resource.SuccessNoopResource;
+            }
+            Thread.Sleep(1000); // Wait 1s before rechecking.
+        }
+    }
+}
+```
+
+### Resource limiter wrapper with reason phrases
+
+```c#
+class ResourceLimiterWrapperReasonPhrase : ResourceLimiter
+{
+    private string _failureReasonPhrase;
+    private ResourceLimiter _innerLimiter;
+
+    public ResourceLimiterWrapperReasonPhrase(string reasonPhrase, ResourceLimiter innerLimiter)
+    {
+        _failureReasonPhrase = reasonPhrase;
+        _innerLimiter = innerLimiter; // wrapping to reduce verbosity in this sample.
+    }
+
+    public long EstimatedCount { get; } => _innerLimiter.EstimatedCount;
+
+    public Resource Acquire(long requestedCount)
+    {
+        var innerResource = _inner.Acquire(requestedCount)
+        if (innerResource.IsAcquired)
+        {
+            return innerResource;
+        }
+
+        var resource = new Resource(true, _failureReasonPhrase /*wrap inner state?*/, _ => innerResource.Dispose());
+        return false;
+    }
+
+    public ValueTask<Resource> AcquireAsync(long requestedCount, CancellationToken cancellationToken = default)
+        => _inner.AcquireAsync(requestedCount, cancellationToken);
+}
+
+// Usage scenario
+
+// API Controller:
+[ApiController]
+[RequestLimit(new ResourceLimiterWrapperReasonPhrase("Too many booking requests", new RateLimiter(requestPerSecond: 100)))]
+public class BookingController : ControllerBase
+{
+    [RequestLimit(new ResourceLimiterWrapperReasonPhrase("Too many hotel booking requests", new RateLimiter(requestPerSecond: 5)))]
```

Comment:
You can't new up things in attributes.

Reply:
Doh, should have verified that this compiles instead coding in a md file.